### ISSUE TITLE
MonitoredResourceBuilder gives incomplete information on GKE by default

### ISSUE THREAD

**Issue by jskeet** [googleapis/gax-dotnet#305]
I'm trying to use MonitoredResourceBuilder on GKE, with a cluster built in a pretty vanilla fashion.

After copying the code from MonitoredResourceBuilder into my code so I can add logging, I can see that I have URLs for `urlNamespace` and `urlPod` to try to get the information, and I also have an access token, but the HTTP request fails with a 403 error.

I'm looking into why...

**Comment by jskeet** [comment]
Okay, I've managed to get this working by changing my cluster configuration:

> Legacy authorisation: Enabled

Ideally we wouldn't require that though...

**Comment by jskeet** [comment]
It looks like "working" is actually a massive overstatement - now `ConfigureGoogleDiagnosticsServices` is failing when it adds exception logging. This code fails:

```csharp
services.AddGoogleExceptionLogging(options =>
{
    options.ProjectId = projectId;
    options.ServiceName = Project.GetAndCheckServiceName(serviceName, monitoredResource);
    options.Version = Project.GetAndCheckServiceVersion(serviceVersion, monitoredResource);
});
```

... because there's no service name. (There's no version, either.) It's not immediately clear to me what those values should be for a GKE application. Continuing to investigate.

**Comment by jskeet** [comment]
When logging `Platform.Instance()`, it's the ContainerName property that's missing unless legacy authorization is enabled.

**Comment by jskeet** [comment]
Looking into this again, the GKE details have everything except the container name, as far as I can see. That's not great, but it's not terrible.

**Comment by jskeet** [comment]
Hmm. Now even with legacy autorization enabled, I'm not getting JSON from the pod - the URL that's failing is https://10.11.240.1:443/api/v1/namespaces/default, with an error of "Unable to read data from the transport connection". Odd.

**Comment by jskeet** [comment]
It's recommended that we use client libraries to access the Kubernetes API. Will try with https://github.com/kubernetes-client/csharp

**Comment by jskeet** [comment]
Odd - we can at least list the namespaces with the client library, so we can get to the API (I think). Trying to work out what the difference is...

**Comment by jskeet** [comment]
And now it's working again. I suspect it takes a while for legacy authorization to *actually* take effect... will try disabling it and see what happens.

**Comment by jskeet** [comment]
Okay, the client fails when legacy authorization is disabled as well. Fundamentally, it looks like in the OOB non-legacy environment, we can't get this information, by design. We get it when we *can* get it.

Currently:

- Logging, error-reporting etc work out of the box
- The container name isn't populated in the monitored resources, but it looks like there's no generally-feasible way of doing that

Basically this bug is infeasible to fix any further, so I'm going to close it.

**Comment by shaneqld** [comment]
Apologies for commenting on an old issue (and interrupting your monologue :)

I had the same issue where containerName was not set. This was caused by my Istio mTLS policy preventing access to kube api server, in addition to my GKE service account not have enough permissions.

I note the code that retrieves the containerName runs this Regex over the contents of /proc/self/mountinfo:

```cs
var regex = new Regex($"/var/lib/kubelet/pods/{podUid}/containers/([^/]+)/.*/dev/termination-log");
```

This had no matches, as podUid could not be determined from the kube api server.

One cat and grep later (within a gke container), I noted that there was only one possible match for this regex in /proc/self/mountinto (when substituting `{podUid}` with `[^/]+`).

As pods and containers have their own file system, might it be suitable to disregard the podUid when determining the containerName (even as a fallback)? This should allow the determination of containerName even with no kube api server access.

Cheers,
Shane.

**Comment by jskeet** [comment]
@shaneqld: Given that the more general problem covered by this issue wouldn't be completely handled by that, could you file it as a separate feature request? I think it does have significant potential, but I wouldn't want to conflate the two.
